test/e2e/external_oidc_test.go

```diff
@@ -0,0 +1,461 @@
+package e2e
+
+import (
+	"context"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+	"testing"
+	"time"
+
+	configv1 "github.com/openshift/api/config/v1"
+	configclient "github.com/openshift/client-go/config/clientset/versioned"
+	operatorversionedclient "github.com/openshift/client-go/operator/clientset/versioned"
+	test "github.com/openshift/cluster-authentication-operator/test/library"
+	"github.com/stretchr/testify/require"
+
+	authenticationv1 "k8s.io/api/authentication/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/apiserver/pkg/storage/names"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+)
+
+const (
+	// TODO: this duplicates the constant defined here: https://github.com/openshift/api/blob/ce6b9c8ad5e785e337869b3bf0f71deeff7bf0eb/features/features.go#L428
+	// until we resolve a dependency breaking compatibility and are able to reference it directly
+	featureGateExternalOIDC = configv1.FeatureGateName("ExternalOIDC")
+
+	oidcClientId      = "admin-cli"
+	oidcAudience      = "openshift-aud"
+	oidcGroupsClaim   = "groups"
+	oidcUsernameClaim = "email"
+
+	kasNamespace = "openshift-kube-apiserver"
+
+	// set if a specific CA bundle is needed for the OIDC provider
+	oidcCABundleConfigMap          = ""
+	oidcCABundleConfigMapNamespace = ""
+)
+
+type testClient struct {
+	t *testing.T
+
+	kubeConfig           *rest.Config
+	kubeClient           *kubernetes.Clientset
+	configClient         *configclient.Clientset
+	operatorConfigClient *operatorversionedclient.Clientset
+}
+
+type oidcAuthResponse struct {
+	AccessToken      string `json:"access_token"`
+	ExpiresIn        int    `json:"expires_in"`
+	RefreshToken     string `json:"refresh_token"`
+	RefreshExpiresIn int    `json:"refresh_expires_in"`
+	TokenType        string `json:"token_type"`
+	IdToken          string `json:"id_token"`
+	NotBeforePolicy  int    `json:"not_before_policy"`
+	SessionState     string `json:"session_state"`
+	Scope            string `json:"scope"`
```

Comment:
Nit :) , why does the struct type use first-character lower case while the struct member uses first-character upper case?

Reply:
The type is lowercase as there's no need to export it to other packages. The members are exported (first-character upper case) because this is needed in order to be able to deserialize JSON data into the struct using `json.Unmarshal()`.